### PR TITLE
Scanner for jekyll projects

### PR DESCRIPTION
**Pull request by vgaidarji**
Fixes https://discuss.bitrise.io/t/create-scanner-for-blog-type-of-projects-static-web-sites/83.

### What has been done
- Introduced a scanner for Jekyll projects
- Enhanced README documentation

### How to test
- Integration test was added
- I've performed a test on various Jekyll projects from Github (https://github.com/programminghistorian/jekyll, https://github.com/barryclark/jekyll-now, 
https://github.com/vgaidarji/jekyll-uno.git).
- Tested on a project (locally) with `_posts` folder only and with `_layouts` folder only. Since at least one of these folders is mandatory.
- Tested the generated config YML from Jekyll scanner on bitrise.io on a test project (https://github.com/barryclark/jekyll-uno).

❔should be the test project used in integration test forked into bitrise organization and then forked version used in test?

### TODO
- [x] create open source sample project for scanner testing
- [x] generate and add `bitrise.yml` to sample project
- [ ] use new sample project in `jekyll_test.go`
- [ ] move test project to https://github.com/bitrise-samples/sample-apps-jekyll
- [ ] add dedicated jekyll build step
- [x] expose the output of the build into an output env var
- [ ] add link to an official guide used to build the scanner
- [ ] recommend the default bitrise stack for jekyll project types

### CODE REVIEW

**Comment by vgaidarji** [comment]
⚠️ `WIP`: This one is almost finished. I just want to perform some more tests before it can be reviewed/merged. I'll ping here once I finish additional tests.

**Comment by vgaidarji** [comment]
Performed additional tests on `jekyll-uno` and `jekyll-now` projects. Ready be reviewed/merged.
I have no access to ci job log and can't see the error. What's wrong with the build and what should I fix?

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ _tmp/
 _bin/
 .bitrise*
 .gows*
+.idea
```
Comment (vgaidarji) [inline .gitignore]: I'm not sure everyone uses GoLand, but if so this one can be removed.

**Comment by trapacska** [comment]
Hey @vgaidarji !

I really like your solution in this PR. However we think that "step logic" (what is currently a script step with the jekyllInitialBuildCommand command) is not acceptable in the scanner. If anything changes with the step logic then it needs to be maintained in the scanner and not in a step.

The second thing is that now both the primary and the deploy workflows are doing the same command. The main reason of this separation is to have testing and deploying in different workflows. Please update these workflows to have a test(primary) which will be ran for each code push(CI) etc..., and a secondary(deploy) workflow that deploys the project. 

For both cases individual steps would be great for these cases. 
For example:
primary:
  - ...
  - ...
  - jekyll-test: run tests and generate reports of it
  - deploy-to-bitrise-io: deploy test reports to the build artifacts section

deploy:
  - ...
  - ...
  - jekyll-test
  - jekyll-build: builds the project prepares the built product to be deployed
  - deploy-to-bitrise-io: deploy stuff to the build artifacts section

After this is done, you can wire in the new steps easily in the scanner you just created.

Thanks in advance!
🚀

**Comment by vgaidarji** [comment]
@trapacska thank you. I just need to find some time to review/implement changes requested. Will be happy to do so asap.

**Comment by vgaidarji** [comment]
@trapacska I'm sorry but I didn't quite get your suggestion here
> I really like your solution in this PR. However we think that "step logic" (what is currently a script step with the jekyllInitialBuildCommand command) is not acceptable in the scanner. If anything changes with the step logic then it needs to be maintained in the scanner and not in a step.

I saw other scanners using step approach too ¯\_(ツ)_/¯.
I'll be happy to adjust whatever needed, just don't know what exactly. 

----- 

> The second thing is that now both the primary and the deploy workflows are doing the same command. The main reason of this separation is to have testing and deploying in different workflows. Please update these workflows to have a test(primary) which will be ran for each code push(CI) etc..., and a secondary(deploy) workflow that deploys the project.

Even though I understand the rationale behind this request, there's no other way of doing things here. Reason: [Jekyll](https://github.com/jekyll/jekyll) projects have no tests and there's no standard way of deploying Jekyll based sites.
In my case I definitely have tests for my Jekyll-based website (Rubocop + HtmlProofer), but we cannot expect everyone having similar configurations. Because of that we cannot standardize the step. What we can do here (💡 I need your advice here) is rename steps names so that they show the intention (one is for CI, another is for CI and deploy)?

For default Jekyll projects the only integration which could happen is only trying to build the site and see if there're no errors.

So it will be:
- `Install dependencies & build` (nothing to push to bitrise or maybe build folder with generated site)
- `Install dependencies & build & deploy` (pushes generated site to bitrise)
So steps will be the same anyway, but only step titles will be different.

**Comment by trapacska** [comment]
Hey @vgaidarji !

> I saw other scanners using step approach too ¯_(ツ)_/¯.

Those are empty/default script steps, to prompt users that they can run their own bash scripts as well. 🙂 

In case there is no tasks for deploying the generated stuff please remove the deploy workflow then. One primary workflow to run the tests could be enough, so jekyll projects can have a simple CI as well.

As soon as you had the deploy workflow removed, I will push this through the team to release the scanner. (also requires some other tasks to be done)

Thanks for the feedbacks.

**Comment by vgaidarji** [comment]
@trapacska done. Please let me know if anything else should be adjusted.
Also, I had to update the Go lint path in 0021128.

----

[In tests](https://github.com/bitrise-core/bitrise-init/pull/116/files#diff-04b4e36eb673fc8488890c74c9c6997aR24) I'm using fork of [jekyll-uno](https://github.com/joshgerdes/jekyll-uno) project. Will you clone `jekyll-uno` repo to `bitrise-samples` as it was done in other tests (e.g. [bitrise-samples/sample-apps-android-sdk22](https://github.com/bitrise-samples/sample-apps-android-sdk22.git))?

**Comment by trapacska** [comment]
Hey @vgaidarji !

We are working on a guide for introducing a new platform support. This way we can avoid unwanted cycles of the PR review. It will take some more time, sorry about that.

**Comment by vgaidarji** [comment]
@trapacska no problem, thanks for the update 👍 
But I would like to fix the build. Maybe you can invite external contributors to scanner builds on bitrise with `ready-only permission`? Otherwise it's not clear why build fails and what to fix 🙃

**Comment by BirmacherAkos** [comment]
Hi @vgaidarji  !

It seems there is a problem in one of our integration tests.
The Tooling Team will fix it soon. 🔧 

The problem is not in your modification, so you don't need to worried about it.

**Comment by godrei** [comment]
Hi @vgaidarji,
really sorry about this long-running pr, this is our first pr for a new scanner and it turned out that we did not clarify a couple of things around this project.

Finally, we managed to create the first version of this project's contribution guide: https://devcenter.bitrise.io/bitrise-cli/creating-your-own-bitrise-project-scanner/

Based on the guide some more things are missing:

1. could you please set up an open source project, which we can use to test the new scanner?

It should include:

- a README.md file (including tool versions required for updating, building and testing this project)
- a bitrise.yml file that is generated by the upcoming scanner

The project, you use for testing would be a great candidate for this: https://github.com/vgaidarji/jekyll-uno
Once the sample project is prepared i would ask you to migrate it to our samples org: https://github.com/bitrise-samples to let us contribute to the project if needed (i would add you to the samples org as a contributor).

2. please create the jekyll build step, as if our scanner generates a starter workflow including script step it is hard to deliver an updates for the logic. Also please expose the output of the build into an output env var.

3. please link an official guide you used to build this scanner
4. please recommend the default stack for this project types

**Comment by vgaidarji** [comment]
@godrei thank you. All requested changes make sense 👍  I'll try to find some free time in the near future and finish the PR.

**Comment by godrei** [comment]
@vgaidarji thanks for your patience! Let me know if i can help you anything.

**Comment by vgaidarji** [comment]
@godrei pinged you in bitrise Slack since I need some clarifications around dedicated jekyll build step.

**Pull request closed without merging**